UI Animations

Banana090 opened the pull request:
Yet another "Crazy Visual Update" everyone will doubt about. Made additional argument for DoButton_MenuTab to make button animate by upscaling it on hover/active state. That's all for now. It's first step in adding some light animations to UI. Making this PR so you can start your favorite discussions about new visuals and so on.

No screenshots, cuz that's animated changes. Look for video on Discord:
[Discord Message Link](https://discordapp.com/channels/252358080522747904/295908390956433410/770091824752099328)
[Download Video](https://cdn.discordapp.com/attachments/295908390956433410/770091820683624458/2020-10-26_04-08-51.mp4)

**Scroll this PR discussion down to see updated videos!!!**

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [?] Tested in combination with possibly related configuration options
- [?] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [?] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

l-ouis commented:
looks nice but text grows too big and doesn't look centered either.

is the upscale linear / smooth / ? btw

Banana090 commented:
@l-ouis it's linear, but smooth xD. Text is aligned by Do_Label code and considering size changes, so it should be centered, but yes, it grows. I can disable this behavior, we'll see

TsFreddie commented on src/game/client/components/menus.cpp:
```diff
@@ -212,18 +241,21 @@ int CMenus::DoButton_MenuTab(const void *pID, const char *pText, int Checked, co
 	CUIRect Temp;
-	pRect->HMargin(2.0f, &Temp);
+	Rect.HMargin(2.0f, &Temp);
 	UI()->DoLabel(&Temp, pText, Temp.h * ms_FontmodHeight, 0, -1, AlignVertically);
```
If you want to scale text. "DoLabel" needs to be changed so you scale the rendered texture instead of trying to render glyphs with new fontsize. It is possible you are spamming multiple redundant glyphs onto the atlas texture under some resolution. Plus pixel alignment in textrender might render the animation shaky in lower resolution.

Banana090 replied:
Not changing text size anymore

Banana090 commented:
I've updated links to video, because I made some changes, look at this now please:
[Discord Message](https://discordapp.com/channels/252358080522747904/295908390956433410/770121606219169822)
[Download Video](https://cdn.discordapp.com/attachments/295908390956433410/770121602355822642/2020-10-26_06-07-44.mp4)

def- commented:
Only having watched the video, I surprisingly like the look, even though I'm generally against animations.

heinrich5991 commented:
I'd be fine with it.

Jupeyy commented on src/game/client/components/menus.cpp:
```diff
@@ -196,14 +216,44 @@ void CMenus::DoButton_KeySelect(const void *pID, const char *pText, int Checked,
+
+	if(pAnimator != NULL)
+	{
+		Time = time();
```
we have time_get_microseconds

Banana090 replied:
no thx, for what reason? Remove 1 extra zero from division? For me it's much clearer to divide by million, not 100.000

Jupeyy commented on src/game/client/components/menus.cpp:
```diff
@@ -196,14 +216,44 @@ void CMenus::DoButton_KeySelect(const void *pID, const char *pText, int Checked,
+	{
+		Time = time();
+
+		if(pAnimator->m_Time + 1000000.f < Time)
```
No need to use floats, they just create inconsitency

Banana090 replied:
think more please

def- replied:
> think more please

Try to be nicer in your communication. Also explain why you actually use a float here instead of your snarky comment. I agree with Jupeyy.

Jupeyy commented on src/game/client/ui.h:
```diff
@@ -97,6 +97,21 @@ class CUIRect
+	bool m_ScaleLabel;
+	bool m_RepositionLabel;
+
+	float m_Time;
```
Here again

Jupeyy reviewed (changes requested):
Don't really like that the browser tabs are round now too.
A step backwards imo

Jupeyy commented on src/game/client/ui.h:
```diff
@@ -103,7 +103,7 @@ struct SUIAnimator
 	bool m_RepositionLabel;
 
-	float m_Time;
+	long long m_Time;
```
why not just int64, just include system.h or whatever its required

edg-l commented:
You don't take into account other languages use longer words, they will be now more out of place than before.

The tabs being round is ugly.

Banana090 commented:
@edg-l the text size does not change at all...

edg-l commented:
Your animation indeed makes the text bigger.

Banana090 commented:
What are we waiting for with this PR? What should I do now for it to be merged? Turned out my next PRs rely on this one

Jupeyy commented:
1. You didnt resolve the requested changes
2. We also have other prs and problems, we cannot add 1 trillion features without testing them
3. You have a conflict that must be resolved

Banana090 commented:
Resolved conflict, changed `long long` to `int64`, included system.h
fixed ultra-rare bug with type overflow so if you open menu just in time (happened every 36 minutes for once) all animations would play once simultaneously.

Jupeyy commented on src/game/client/components/menus.cpp:
```diff
@@ -94,6 +94,26 @@ CMenus::CMenus()
 	m_ServerProcess.Process = 0;
 	m_ServerProcess.Initialized = false;
+
+	for(SUIAnimator &animator : m_aAnimatorsSettingsTab)
```
Capital A

Jupeyy commented on src/game/client/components/menus.cpp:
```diff
@@ -94,6 +94,26 @@ CMenus::CMenus()
+		animator.m_RepositionLabel = true;
+	}
+
+	for(SUIAnimator &animator : m_aAnimatorsBigPage)
```
Capital A

Jupeyy commented on src/game/client/components/menus.cpp:
```diff
@@ -94,6 +94,26 @@ CMenus::CMenus()
+		animator.m_HOffset = 5.0f;
+	}
+
+	for(SUIAnimator &animator : m_aAnimatorsSmallPage)
```
Capital A,

Also, you are only initializing a few values of these variables, and don't explicitly have a constructor for `SUIAnimator`

Jupeyy reviewed (commented):
You could try and see if your IDE or text editor has a clang-format extension or plugin or whatever, so you don't always run into these style issues

heinrich5991 commented:
@Banana090 Are you still interested in updating this?
